designs/node-repair.md

```diff
@@ -0,0 +1,182 @@
+2. If the node is determined to unhealthy, a status condition will be added to the NodeClaim
+3. Disruption controller will force terminate the node claim immediately 
+
+```
+apiVersion: karpenter.sh/v1
+kind: NodeClaim
+metadata:
+  ...
+status:
+  conditions:
+    - lastTransitionTime: "2024-11-04T21:06:06Z"
+      status: "True"
+      reason: <cloud provider returned UnhealthyReason>
+      type: Unhealthy
+    ...
+```
+
+### Option 3: Cloud provider injected unhealthy condition 
+
+```
+type UnhealthyType metav1.ConditionType 
+type UnhealthyRepairAfter time.Duration
+
+// UnhealthyConditions are condition to watch for on nodes inside of the cluster
+var UnhealthyConditions = map[UnhealthyType]UnhealthyRepairAfter{
+    "Ready": "30m"
+    "DiskPressure": "10m"
+    ...
+}
+```
+
+Cloud provider hydrate a `UnhealthyConditions` map that will hold the condition to track on nodes. The map will track the condition and `UnhealthyRepairAfter` will dictate the length of time the node will be inside the cluster before action is taken against it. Node health validation will follow as such:
+
+1. A diagnostic agent will add a status condition on to a node 
+2. Karpenter will reconcile on nodes and identify unhealthy nodes defined by `UnhealthyConditions` map
+3. Unhealthy condition is added to the NodeClaim
+4. Disruption controller will forcefully terminate the the NodeClaim once the node has been in an unhealthy state for the duration specified by `UnhealthyConditions` map
+
+## Recommended Solution
+
+The recommended approach is for cloud provider interface should expose a layer of configuration for Karpenter to monitor nodes and take action. I propose we implement option one. Cloud provider can define any set of status conditions that are tracked on the node. An interface approach allows for easy extensibility. Option 2 has the approach of being over generalized as we don’t have use-cases to show the need. All solutions outlined below will be under a `NodeRepair=true`  feature flag to indicate that these feature are not stable.
+
+### Forceful termination
+
+For a first iteration approach Karpenter will implement force termination. Today, graceful termination in Karpenter will attempt to wait for pod to be fully drained on a node and all volume attachment to be deleted from a node. This raises the problem that during graceful termination node can be stuck terminating as pod eviction or volume detachment may be broken. In these cases, users will need to take manual action against the node. **For the Alpha implementation, the recommendation will be to non-graceful force termination nodes. Furthermore, unhealthy nodes will not respect the customer configured terminationGracePeriod.**
+
+## Future consideration 
+
+The recommended solution includes a limited set of features that will improve the customer experience. There are additional feature we will like to include once more user date can help drive the community decision. These include:
```

Comment:
You may not need this statement here -- you could probably shorten this and just say something similar to "Here's what we are considering after releasing this feature: ..."